### PR TITLE
Add a guide on authoring tpgtools new resources

### PR DESCRIPTION
**Pull request by slevenick**
Let me know if you have feedback!

Fixes: https://github.com/hashicorp/terraform-provider-google/issues/10544

If this PR is for Terraform, I acknowledge that I have:

- [ ] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [ ] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-downstream-tools), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [ ] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/third_party/terraform/tests) (for handwritten resources or update tests).
- [ ] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/master/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [ ] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/master/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:none

```

### CODE REVIEW

File: tpgtools/README.md [hunk 1]
```diff
@@ -176,3 +176,33 @@ doc_hide:
 test_hide:
   - basic_workload.yaml
 ```
+
+## New Resource Guide (tpgtools only)
+
+This guide is written to document the process for adding a resource to the Google Terraform Provider (TPG) after it has been added to the [DCL](https://github.com/GoogleCloudPlatform/declarative-resource-client-library).
+
+### Adding Resource Overrides
```
Comment (rileykarson) [inline tpgtools/README.md:184]: For this + the header below: I'd recommend assuming that the override file will _always_ be blank, explain briefly what an override is, and recommend that folks reach out to an experienced team member if they believe an override will be needed for their resource(s). To cover brownfield resources, I'd leave them out of this flow entirely and add them as an appendix afterwards- "those were the guidelines for new resources, here's how a brownfield resource is different"

Motivation(s):
* This keeps things simple for folks that are making their first contribution. It isn't true that you never need overrides yet, but we need to make it true.
* Deferring to an experienced engineer will mean they can help solve the root cause that needed an override in the first place.
* This is in contrast to the MMv1 field guide, where I explain some MMv1 overrides- using them is pretty normal, even for simple contributions.
* We expect most brownfield conversions to be done for our team (or another experienced member of a sister team) so having the information out of a new dev's critical path means there's less information overload

File: tpgtools/README.md [hunk 2]
```diff
@@ -176,3 +176,33 @@ doc_hide:
...
+
+#### What should I add to the overrides file?
+
+For the first pass try nothing. If behavior is needed to change the resource's behavior in Terraform from the default DCL resource, then you can add these at a later date. Generally no overrides will be needed for a new resource, but they will be needed for resources that exist in mmv1 that are being converted to use tpgtools.
+
+### Adding Samples
+
+See [guide above](#samples) for adding samples.
+
+### Adding Additional Information to Handwritten/MMv1 Generated Files
```
Comment (rileykarson) [inline tpgtools/README.md:196]: nit: "Additional Information..." feels optional. "Registering the Resource", maybe?

File: tpgtools/README.md [hunk 3]
```diff
@@ -176,3 +176,33 @@ doc_hide:
...
+
+### Adding Resource Overrides
+
+Every resource added via tpgtools needs an override file for every version it is available at. This file may be empty, but must exist. A resource available at GA (TPG) must also exist at beta (TPGB) and needs a corresponding override file at beta. These override files are often identical between versions. This file should exist at tpgtools/overrides/$PRODUCT_NAME/$VERSION/$RESOURCE.yaml. For example, [this override](https://github.com/GoogleCloudPlatform/magic-modules/blob/master/tpgtools/overrides/assuredworkloads/beta/workload.yaml) exists for the product assuredworkloads resource workload at beta version.
+
+#### What should I add to the overrides file?
+
+For the first pass try nothing. If behavior is needed to change the resource's behavior in Terraform from the default DCL resource, then you can add these at a later date. Generally no overrides will be needed for a new resource, but they will be needed for resources that exist in mmv1 that are being converted to use tpgtools.
+
+### Adding Samples
```
[the remark this answers is not in the document]
Comment (rileykarson) [reply tpgtools/README.md:192]: @ndmckinley also recently revamped how we pull in substitutions, I think less work is needed on those than is written here.

File: tpgtools/README.md [hunk 4]
```diff
@@ -176,3 +176,33 @@ doc_hide:
...
+
+#### Adding to provider.go
+
+Add the resource definition within the `ResourceMapWithErrors` map. There is currently a block of tpgtools resources, so adding to that block is preferred. If this resource is available only at a specific version (beta, private) then add version tags around the resource definition.
+
+Additionally you may need to add a few line to provider.go for the endpoint if this resource is in a new product that doesn't have any other resources currently in the provider. Find the correct lines in the diff on `provider_dcl_endpoints.go` within the provider once it has been generated with the new resource.
+
+### Adding Documentation
```
Comment (rileykarson) [inline tpgtools/README.md:206]: We may want a section about writing handwritten tests as well! Copying the DCL samples may not result in the level of testing we want, or we may need customised tests for some resources.

File: tpgtools/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (ndmckinley) [reply tpgtools/README.md:192]: Yeah, this should Just Work for new versions of the DCL now.  There's some no-op cleanup to do with deduplicating variables as it stands, but tpgtools handles it fine.

File: tpgtools/README.md [hunk 5]
```diff
@@ -51,39 +51,24 @@ To generate accessory code such as `serializarion`, you can specify the `--mode`
...
 
-We will autoingest samples from the dcl, however we currently must
-manually fill the substitutions for these samples.
+Start by copying the relevant samples from the DCL for your new resource. These will be added to
+the [tpgtools/api folder](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/tpgtools/api) under the relevant product.
+These samples can be found under the samples/ folder within the DCL for the resource being added. For example, assured
+workloads can be found [here](https://github.com/GoogleCloudPlatform/declarative-resource-client-library/tree/main/services/google/assuredworkloads/samples).
 
 You may need to re-serialize `serialization.go` if you are adding newer resources.
 To do this
```
Comment (rileykarson) [inline tpgtools/README.md:74]: Nit: we can use `make upgrade-dcl` below rather than `go get` directly.

File: tpgtools/README.md [hunk 6]
```diff
@@ -51,39 +51,24 @@ To generate accessory code such as `serializarion`, you can specify the `--mode`
...
+Adding samples is essential for generating tests and documentation.
 
-We will autoingest samples from the dcl, however we currently must
-manually fill the substitutions for these samples.
+Start by copying the relevant samples from the DCL for your new resource. These will be added to
+the [tpgtools/api folder](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/tpgtools/api) under the relevant product.
+These samples can be found under the samples/ folder within the DCL for the resource being added. For example, assured
+workloads can be found [here](https://github.com/GoogleCloudPlatform/declarative-resource-client-library/tree/main/services/google/assuredworkloads/samples).
 
 You may need to re-serialize `serialization.go` if you are adding newer resources.
```
Comment (rileykarson) [inline tpgtools/README.md:73]: nit: Given this is a new resource guide, we can omit this condition.

File: tpgtools/README.md [hunk 7]
```diff
@@ -176,3 +114,50 @@ doc_hide:
...
   - basic_workload.yaml
 ```
+
+### Registering New Resources
+
+There are a few files within mmv1 that use Ruby templates to generate shared files for the provider. One of these is provider.go (generated by [provider.go.erb](https://github.com/GoogleCloudPlatform/magic-modules/blob/master/mmv1/third_party/terraform/utils/provider.go.erb)) which contains the map of resources to their names within the provider. Currently these need to be updated by hand to add a new resource. There are a couple of other places that need to be modified as well to add endpoint information if the resource being added is part of a new product.
+
+#### Adding to provider.go
+
+Add the resource definition within the `ResourceMapWithErrors` map. There is currently a block of tpgtools resources, so adding to that block is preferred. If this resource is available only at a specific version (beta, private) then add version tags around the resource definition.
```
Comment (rileykarson) [inline tpgtools/README.md:124]: nit: Consider using a code block to show a sample diff!

**Review by rileykarson — APPROVED** [review]

**Pull request merged**